File: backend/models/account.js
"use strict";

module.exports = (sequelize, DataTypes) => {
  const Account = sequelize.define(
    "Account",
    {
      id: {
        type: DataTypes.INTEGER.UNSIGNED,
        primaryKey: true,
        autoIncrement: true,
        comment: "PK",
      },
      accountNumber: {
        type: DataTypes.STRING(32),
        allowNull: false,
        comment: "Номер особового рахунку",
      },
      legalName: {
        type: DataTypes.STRING(255),
        allowNull: false,
        comment: "Повна юридична назва установи",
      },
      edrpou: {
        type: DataTypes.STRING(8),
        allowNull: false,
        comment: "Код ЄДРПОУ",
      },
      rozporiadNumber: {
        type: DataTypes.STRING(32),
        allowNull: true,
        comment: "Відомчий код",
      },
      additionalInfo: {
        type: DataTypes.TEXT,
        allowNull: true,
        comment: "Додаткова інформація",
      },
      subordinationId: {
        type: DataTypes.INTEGER.UNSIGNED,
        allowNull: false,
        comment: "Підпорядкованість",
      },
    },
    {
      tableName: "Account",
      timestamps: true,
      comment: "Особові рахунки",
    }
  );

  Account.associate = (models) => {
    // належить довіднику підпорядкування
    if (models.Subordination)
      Account.belongsTo(models.Subordination, {
        foreignKey: "subordinationId",
      });

    // використовується у різних документах/реєстрах як FK
    if (models.EstimateLine)
      Account.hasMany(models.EstimateLine, { foreignKey: "accountId" });

    if (models.ProposalLine)
      Account.hasMany(models.ProposalLine, { foreignKey: "accountId" });

    if (models.BudgetExecution)
      Account.hasMany(models.BudgetExecution, { foreignKey: "accountId" });

    if (models.Request)
      Account.hasMany(models.Request, { foreignKey: "accountId" });

    if (models.RequestRegister)
      Account.hasMany(models.RequestRegister, { foreignKey: "accountId" });
  };

  return Account;
};

File: fin/lib/core/models/account.dart
class Account {
  final String? id;
  final String accountNumber;
  final String rozporiadNumber;
  final String legalName;
  final String edrpou;
  final String? subordination;
  final String additionalInfo;

  Account({
    this.id,
    required this.accountNumber,
    required this.rozporiadNumber,
    required this.legalName,
    required this.edrpou,
    this.subordination,
    this.additionalInfo = '',
  });

  factory Account.fromJson(Map<String, dynamic> j) => Account(
    id: j['id'] as String?,
    accountNumber: j['accountNumber'] ?? '',
    rozporiadNumber: j['rozporiadNumber'] ?? '',
    legalName: j['legalName'] ?? '',
    edrpou: j['edrpou'] ?? '',
    subordination: j['subordination'],
    additionalInfo: j['additionalInfo'] ?? '',
  );

  Map<String, dynamic> toJson() {
    final m = {
      'accountNumber': accountNumber,
      'rozporiadNumber': rozporiadNumber,
      'legalName': legalName,
      'edrpou': edrpou,
      'subordination': subordination,
      'additionalInfo': additionalInfo,
    };
    if (id != null) m['id'] = id;
    return m;
  }
}
